design/tool.behavior.md

```diff
@@ -0,0 +1,95 @@
+#  Tool Behavior
+
+The tool behavior allows the user to assign this behavior to an actor so that the actor can be held and
+used.
+
+## Architecture
+
+Create a new behavior that is an `ToolBehavior` for being able to add to an actor, that when added will
+automatically enable grabbable on that actor.  This behavior will then allow the user to execute a primary 
+action to begin recording the transform changes over time.
+
+``` ts
+/**
+ * Tool behavior class containing the target behavior actions.
+ */
+export class ToolBehavior extends TargetBehavior {
+	private _holding: DiscreteAction = new DiscreteAction();
+	private _using: DiscreteAction = new DiscreteAction();
+
+	/** @inheritdoc */
+	public get behaviorType(): BehaviorType { return 'tool'; }
+
+	public get holding() { return this._holding; }
+	public get using() { return this._using; }
+
+	/**
```

Comment:
I can imagine many cases where it would be useful to get an `Actor`'s transform stream that are completely unrelated to `ToolBehavior`. Because of this, I don't think `onUsingPathReceived` belongs here. I propose adding this API to `Actor` instead:
```ts
public onTransformChanged((path: Transform[]) => void) {
   // Transform changed callback registration
   // Auto-subscribe to transform changes
}
public offTransformChanged((path: Transform[]) => void) {
   // Transform changed callback de-registration
   // Auto-unsubscribe to transform changes if this was the last registered callback
}
```

Reply:
If you want this as an all up addition to actor, I am not opposed to that.  I was recommending it on the tool itself because it logically makes sense to have that API ties to the tool.  I think I would prefer adding this to the actor and still having an API on the pen that wraps this call on the actor.  It is more intuitive and discoverable to the developer to know they are using a pen tool and can get the use path right from it.

Reply:
Ok then lets put it on Actor. Some thoughts on implementation:
* I think we can make all transform subscriptions high-fidelity, collecting at 10Hz.
* On the client, it seems like we should collect the transforms in FixedUpdate, so that high-end and low-end clients collect the same number of transforms per update.
* Wondering the best way to package these into a payload. A list of transforms could be added to the ActorPatch.

About this API existing here too .. I think this API should be removed from the behavior for two reasons:
* We should try to avoid defining two APIs that do the same thing.
* These functions are too opinionated. They assume to know when you want the transforms and when you don't. It's better to let the app developer write their own wrappers that encode these kids of assumptions.

Reply:
Fair point.  I will remove it.

Reply:
Now that I think about it, this is actually a separate API.  This isn't simply a wrapper around the registration of the transform updates.  There is more to it than that.  We may want additional info to come along through the recording.  Also we want latency to be invisible with something like inking and doing the recording in the app will not enable zero latency for the path generation.  We should have an online discussion about this.

Reply:
> Also we want latency to be invisible with something like inking and doing the recording in the app will not enable zero latency for the path generation.

Can you clarify your meaning here? We've split the inking operation into two parts: a zero-latency "wet ink" effect, and mesh generation which does have latency. Which part are you referring to here?

Additional info like pen angle, pressure, etc would allow for richer curve rendering. Those could be standard brush inputs. I wonder what kinds of data other pen and paint programs encode into their stroke recordings.

Reply:
I think after the call, the latency I am concerned about centers around the recording of the transforms at the time that the draw begins on the client, and not after a trip to the server for the perform action call to be made.  I am updating the review based on the chat we had to handle this latency and still support the transform update API on the actor.